flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/CodeGenerator.scala

```diff
@@ -871,12 +871,24 @@ class CodeGenerator(
       returnType: TypeInformation[_ <: Any],
       resultFieldNames: Seq[String])
     : GeneratedExpression = {
-    val input1AccessExprs = for (i <- 0 until input1.getArity if input1Mapping.contains(i))
-      yield generateInputAccess(input1, input1Term, i, input1Mapping)
+
+    val input1AccessExprs = if (input1.isInstanceOf[PojoTypeInfo[_]]) {
+      for (i <- 0 until input1Mapping.length)
+        yield generateInputAccess(input1, input1Term, i, input1Mapping)
+    } else {
+      for (i <- 0 until input1.getArity if input1Mapping.contains(i))
+        yield generateInputAccess(input1, input1Term, i, input1Mapping)
+    }
 
     val input2AccessExprs = input2 match {
-      case Some(ti) => for (i <- 0 until ti.getArity if input2Mapping.contains(i))
-        yield generateInputAccess(ti, input2Term, i, input2Mapping)
+      case Some(ti) => if (input2.isInstanceOf[PojoTypeInfo[_]]) {
```

Comment:
I think your branch looks good @wuchong. I tried a similar thing but did not change the `generateFieldAccess()` methods because I was afraid of side effects if we use the mapping for any kind types (instead of just POJOs).